addons/payment/controllers/portal.py

```diff
@@ -1,376 +1,437 @@
-# -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
-import hashlib
-import hmac
-import logging
-from unicodedata import normalize
-import psycopg2
+import urllib.parse
 import werkzeug
 
-from odoo import http, _
+from odoo import _, http
+from odoo.exceptions import UserError, ValidationError
 from odoo.http import request
-from odoo.osv import expression
-from odoo.tools import DEFAULT_SERVER_DATETIME_FORMAT, consteq, ustr
-from odoo.tools.float_utils import float_repr
-from datetime import datetime, timedelta
 
-
-_logger = logging.getLogger(__name__)
-
-class PaymentProcessing(http.Controller):
-
-    @staticmethod
-    def remove_payment_transaction(transactions):
-        tx_ids_list = request.session.get("__payment_tx_ids__", [])
-        if transactions:
-            for tx in transactions:
-                if tx.id in tx_ids_list:
-                    tx_ids_list.remove(tx.id)
+from odoo.addons.payment import utils as payment_utils
+from odoo.addons.payment.controllers.post_processing import PaymentPostProcessing
+from odoo.addons.portal.controllers import portal
+
+
+class PaymentPortal(portal.CustomerPortal):
+
+    """ This controller contains the foundations for online payments through the portal.
+
+    It allows to complete a full payment flow without the need of going though a document-based flow
+    made available by another module's controller.
+
+    Such controllers should extend this one to gain access to the _create_transaction static method
+    that implements the creation of a transaction before its processing, or to override specific
+    routes and change their behavior globally (e.g. make the /pay route handle sale orders).
+
+    The following routes are exposed:
+    - `/payment/pay` allows for arbitrary payments.
+    - `/my/payment_method` allows the user to create and delete tokens. It's its own `landing_route`
+    - `/payment/transaction` is the `transaction_route` for the standard payment flow. It creates a
+      draft transaction, and return the processing values necessary for the completion of the
+      transaction.
+    - `/payment/confirmation` is the `landing_route` for the standard payment flow. It displays the
+      payment confirmation page to the user when the transaction is validated.
+    - `/payment/validation` is the `validation_route` for the standard payment method validation
+      flow. It redirects the user to `/my/payment_method` in order to display the result and allow
+      to start the flow over.
+    """
+
+    # TODO TBE According to feedback 2353112, this route is used to test credit cards numbers on our
+    #          customers' database. Providers such as Authorize.Net allows passing the IP address of
+    #          the visitor to filter the payment attempts. Shouldn't we design something at the
+    #          level of the framework to prevent such attacks?
```

Comment:
Sure, it is a great idea to pass the customer IP Address, but it can be done on the payment modules, no need to do something in the framework level (too low on the stack?)

Reply:
Done and tested for Adyen and Authorize in 51a7bd0f054